casper/src/main/scala/coop/rchain/casper/ReportingCasper.scala

```diff
@@ -240,88 +173,262 @@ object ReportingCasper {
+    * https://rchain.atlassian.net/browse/RCHAIN-4025, the coop.rchain.casper.util.rholang.RuntimeManagerImpl should
+    * be refactored in the future and the runtimeManager here should be refactored along with it and make
+    * the reportingCasper work with the the refactored runtimeManager. The duplicate codes should be refactored anyway.
+    * The codes here are very ugly now.
+    */
   class ReportingRuntimeManagerImpl[F[_]: Concurrent: Metrics: Span: Log](
       val emptyStateHash: StateHash,
       runtimeContainer: MVar[F, ReportingRuntime[F]]
   ) {
+    import coop.rchain.models.rholang.{implicits => toPar}
+    private val systemDeployConsumeAllPattern =
+      BindPattern(List(toPar(Expr(EVarBody(EVar(Var(FreeVar(0))))))), freeCount = 1)
+    private val emptyContinuation = TaggedContinuation()
+    private def setInvalidBlocks(
+        invalidBlocks: Map[BlockHash, Validator],
+        runtime: ReportingRuntime[F]
+    ) = {
+      val invalidBlocksPar: Par =
+        Par(
+          exprs = Seq(
+            Expr(
+              Expr.ExprInstance.EMapBody(
+                ParMap(SortedParMap(invalidBlocks.map {
+                  case (validator, blockHash) =>
+                    (
+                      RhoType.ByteArray(validator.toByteArray),
+                      RhoType.ByteArray(blockHash.toByteArray)
+                    )
+                }))
+              )
+            )
+          )
+        )
+      runtime.invalidBlocks.setParams(invalidBlocksPar)
+    }
 
     def replayComputeState(startHash: StateHash)(
         terms: Seq[ProcessedDeploy],
-        blockData: BlockData
-    ): F[Either[ReplayFailure, List[(ProcessedDeploy, Seq[ReportingEvent])]]] =
+        sysDeploys: Seq[ProcessedSystemDeploy],
+        blockData: BlockData,
+        invalidBlocks: Map[BlockHash, Validator],
+        isGenesis: Boolean
+    ): F[Either[ReplayFailure, List[(ProcessedDeploy, Seq[Seq[ReportingEvent]])]]] =
       Sync[F].bracket {
         runtimeContainer.take
       } { runtime =>
-        replayDeploys(runtime, startHash, terms, replayDeploy(runtime))
+        for {
+          _ <- runtime.blockData.set(blockData)
+          _ <- setInvalidBlocks(invalidBlocks, runtime)
+          result <- replayDeploys(
+                     runtime,
+                     startHash,
+                     terms,
+                     sysDeploys,
+                     replayDeploy(runtime, !isGenesis)
+//                     replaySystemDeploy(runtime)
+                   )
+        } yield result
       }(runtimeContainer.put)
 
     private def replayDeploys(
         runtime: ReportingRuntime[F],
         startHash: StateHash,
         terms: Seq[ProcessedDeploy],
-        replayDeploy: ProcessedDeploy => F[Either[ReplayFailure, Seq[ReportingEvent]]]
-    ): F[Either[ReplayFailure, List[(ProcessedDeploy, Seq[ReportingEvent])]]] =
+        systemDeploys: Seq[ProcessedSystemDeploy],
```

Comment:
System deploys are not used. Without system deploys REV transactions to charge deploys are not visible in the report. Do we need ticket for that or is it necessary now?

Reply:
No. The preCharge and refund system deploys are all included in the ProcessedDeploy.
The sytemDeploys here are just for closeBlock and slashDeploy. We don't need that now so I didn't put my effort on it. The soft reports above are  the ways the distinguish the precharge , refund and userDeploy.
These https://github.com/rchain/rchain/blob/dev/models/src/main/protobuf/CasperMessage.proto#L148 deploy logs includes precharge , refund and userDeploy logs. `closeBlock` and `slashDeploy` deploy logs are in https://github.com/rchain/rchain/blob/dev/models/src/main/protobuf/CasperMessage.proto#L170. I don't actually like the design which put precharge, refund and userDeploy event logs altogether.

Reply:
You are right, we don't need close block and slash logs. Parameter can be removed altogether.